Improve arrayStableSort using Powersort and binary insertion sort

danlliu opened the pull request:
#### 04c766519a71e85e84f1fbb81b469874553eec96
<pre>
Improve arrayStableSort using Powersort and binary insertion sort
<a href="https://bugs.webkit.org/show_bug.cgi?id=288130">https://bugs.webkit.org/show_bug.cgi?id=288130</a>
<a href="https://rdar.apple.com/145235779">rdar://145235779</a>

Reviewed by Yusuke Suzuki.

Currently, JSC uses mergesort and naive insertion sort to sort a JSArray.
We can improve this by implementing Powersort, and updating our insertion
sort to use binary search to find the insertion position.

tagcloud-SP: average ~3% better (significant with 98% probability)
UniPoker: average ~1% better (significant with 98% probability)

* JSTests/microbenchmarks/sort-custom-comparator.js: Added.
(randArray):
(comparator):
(iteration.i.comparator):
* Source/JavaScriptCore/runtime/StableSort.h:
(JSC::arrayInsertionSort):
(JSC::mergePowersortRuns):
(JSC::arrayStableSort):
(JSC::arrayMerge): Deleted.

Canonical link: <a href="https://commits.webkit.org/290894@main">https://commits.webkit.org/290894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f484b5a4b4d17cb483fb953cfa7cc859e5a3770

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96214 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41957 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11153 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19073 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70087 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27600 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8499 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82637 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50413 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/212 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41096 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84032 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98199 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89979 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18407 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13488 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18663 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78468 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22807 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/158 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11581 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14460 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18407 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23711 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112551 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18128 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32668 "Found 12 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-medium.js.no-llint, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-slow-memory, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.default-wasm, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-slow-memory, wasm.yaml/wasm/stress/repro_1289.js.default-wasm, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +104,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+        unsigned power;
+    };
+
+    WTF::Vector<PowersortStackEntry, 16> powerstack;
```
How about reserving the max size as the capacity? I think it is not so large.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -45,54 +45,46 @@ static ALWAYS_INLINE bool coerceComparatorResultToBoolean(JSGlobalObject* global
-    size_t right = leftEnd;
-    size_t rightEnd = std::min<size_t>(right + width, srcEnd);
-
-    if (right >= rightEnd) {
```
This part is very important optimization (without it, SP3 will regress). Is it ok to remove this?

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -45,54 +45,46 @@ static ALWAYS_INLINE bool coerceComparatorResultToBoolean(JSGlobalObject* global
+    size_t right = srcIndex2;
+    size_t rightEnd = srcEnd2;
 
-    bool result = comparator(src[right], src[right - 1]);
```
Ditto

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+    while (run1.m_end + 1 < numElements && !comparator(from[run1.m_end + 1], from[run1.m_end])) {
         RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
+        ++run1.m_end;
     }
```
Let's write `comparator` calls simpler.
```
while (run1.m_end + 1 < numElements) {
    bool result = comparator(from[run1.m_end + 1], from[run1.m_end]);
    RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
    if (result)
        break;
    ++run1.m_end;
}
```

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+    while (run1.m_end + 1 < numElements && !comparator(from[run1.m_end + 1], from[run1.m_end])) {
+        RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
+        ++run1.m_end;
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+        while (run2.m_end + 1 < numElements && !comparator(from[run2.m_end + 1], from[run2.m_end])) {
+            RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
+            ++run2.m_end;
+        }
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+
+        if (run2.m_end - run2.m_begin < extendRunCutoff) {
+            // If the run is too short, insertion sort a bit
+
```
Remove this empty line.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+        while (run2.m_end + 1 < numElements && !comparator(from[run2.m_end + 1], from[run2.m_end])) {
             RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
+            ++run2.m_end;
         }
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+    if (numElements < extendRunCutoff) {
+        arrayInsertionSort(vm, src.subspan(0, src.size()), comparator);
+        RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope, src);
+        return src;
```
`RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED` is not necessary. Let's do,
```
scope.release();
arrayInsertionSort(vm, src.subspan(0, src.size()), comparator);
return src;
```

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+        // n << 62 must be <= 2^126, so a << 62 must be < 2^127. thus, we don't end up with overflow
+        a <<= 62;
+        b <<= 62;
+        return WTF::clz((a / n) ^ (b / n)) - 64;
```
Does `clz` work with `UInt128`?
Also, WTF:: is not necessary.

danlliu replied:
I confirmed the power calculation still works correctly with UInt128; `clz` seems to work fine.

Constellation commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -112,25 +105,128 @@ static ALWAYS_INLINE void arrayMerge(VM& vm, std::span<ElementType> dst, std::sp
+        WTF::UInt128 n2 = right - middle + 1;
+        // a and b are 2*midpoints of the two ranges, so always within [0, 2n)
+        WTF::UInt128 a = left * 2 + n1;
+        WTF::UInt128 b = middle * 2 + n2;
```
We do not need `WTF::` prefix.

danlliu committed:
Improve arrayStableSort using Powersort and binary insertion sort
https://bugs.webkit.org/show_bug.cgi?id=288130
rdar://145235779

Reviewed by Yusuke Suzuki.

Currently, JSC uses mergesort and naive insertion sort to sort a JSArray.
We can improve this by implementing Powersort, and updating our insertion
sort to use binary search to find the insertion position.

tagcloud-SP: average ~3% better (significant with 98% probability)
UniPoker: average ~1% better (significant with 98% probability)

* JSTests/microbenchmarks/sort-custom-comparator.js: Added.
(randArray):
(comparator):
(iteration.i.comparator):
* Source/JavaScriptCore/runtime/StableSort.h:
(JSC::arrayInsertionSort):
(JSC::mergePowersortRuns):
(JSC::arrayStableSort):
(JSC::arrayMerge): Deleted.

Canonical link: https://commits.webkit.org/290894@main
